pkg/controller/master/resourcequota/resourcequota_controller.go

```diff
@@ -0,0 +1,81 @@
+package resourcequota
+
+import (
+	"errors"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+
+	ctlharvcorev1 "github.com/harvester/harvester/pkg/generated/controllers/core/v1"
+	"github.com/harvester/harvester/pkg/util"
+	rqutils "github.com/harvester/harvester/pkg/util/resourcequota"
+)
+
+type Handler struct {
+	rqs     ctlharvcorev1.ResourceQuotaClient
+	rqCache ctlharvcorev1.ResourceQuotaCache
+}
+
+var errSkipScaling = errors.New("skip scaling")
+
+// When Harvester is scaling resourcequota per vmim, the resourcequota can also be reset by Rancher
+// Luckily Rancher only resets it when Rancher related POD starts, it does not monitor and reset it all the time
+// Harvester controller calculates the resourcequota per Rancher annotation and Harvester annotation
+// If someday Rancher keeps resetting it, this controller needs an enhancement to avoid looping update resourcequota with Rancher
+func (h *Handler) OnResourceQuotaChanged(_ string, rq *corev1.ResourceQuota) (*corev1.ResourceQuota, error) {
+	// not a target resourcequota
+	if rq == nil || rq.DeletionTimestamp != nil || rq.Annotations == nil || rq.Labels == nil || rq.Labels[util.LabelManagementDefaultResourceQuota] != "true" {
+		return rq, nil
+	}
+
+	rqCopy := rq.DeepCopy()
+	update, err := scaleResourceQuotaOnDemand(rqCopy)
+	if err != nil {
+		if errors.Is(err, errSkipScaling) {
+			return rq, nil
+		}
+		return rq, err
+	}
+	if update {
+		return h.rqs.Update(rqCopy)
+	}
+
+	return rq, nil
+}
+
+func scaleResourceQuotaOnDemand(rq *corev1.ResourceQuota) (bool, error) {
+	// below data is only related to rq itself, if error happens and run reconciller, it will fall into error looping
+	update := false
+	// if both CPU and memory have no limits
+	if rqutils.IsEmptyResourceQuota(rq) {
+		logrus.Debugf("resourcequota %s/%s has 0 quota, skip scaling", rq.Namespace, rq.Name)
+		return update, errSkipScaling
+	}
+
+	carq, _ := rq.Annotations[util.CattleAnnotationResourceQuota]
```

Comment:
The annotation is on namespace object, not the RQ object, will update this block of code.